src/share/cxx/Diagnostics.cpp

```diff
@@ -0,0 +1,226 @@
+#include "Diagnostics.hpp"
+
+#include "Context.hpp"
+#include "Elements.hpp"
+#include "HybridVCoord.hpp"
+#include "PhysicalConstants.hpp"
+#include "SimulationParams.hpp"
+#include "TimeLevel.hpp"
+#include "Tracers.hpp"
+
+#include "utilities/SubviewUtils.hpp"
+
+namespace Homme
+{
+
+void Diagnostics::init (const int num_elems, F90Ptr& elem_state_q_ptr, F90Ptr& elem_accum_qvar_ptr,
+                        F90Ptr& elem_accum_qmass_ptr, F90Ptr& elem_accum_q1mass_ptr,
+                        F90Ptr& elem_accum_iener_ptr, F90Ptr& elem_accum_iener_wet_ptr,
+                        F90Ptr& elem_accum_kener_ptr, F90Ptr& elem_accum_pener_ptr)
+{
+  assert (num_elems>0);
+  m_num_elems = num_elems;
+
+  // F90 ptr to array (n1,n2,...,nK,nelemd) can be stuffed directly in an unmanaged view
+  // with scalar type Real*[nK]...[n2][n1] (with runtime dimension nelemd)
+  h_Q      = HostViewUnmanaged<Real*[QSIZE_D][NUM_PHYSICAL_LEV][NP][NP]>(elem_state_q_ptr,num_elems);
+  h_Qvar   = HostViewUnmanaged<Real*[4][QSIZE_D][NP][NP]>(elem_accum_qvar_ptr,num_elems);
+  h_Qmass  = HostViewUnmanaged<Real*[4][QSIZE_D][NP][NP]>(elem_accum_qmass_ptr,num_elems);
+  h_Q1mass = HostViewUnmanaged<Real*   [QSIZE_D][NP][NP]>(elem_accum_q1mass_ptr,num_elems);
+  h_IEner     = HostViewUnmanaged<Real*[4][NP][NP]>(elem_accum_iener_ptr,num_elems);
+  h_IEner_wet = HostViewUnmanaged<Real*   [NP][NP]>(elem_accum_iener_wet_ptr,num_elems);
+  h_KEner     = HostViewUnmanaged<Real*[4][NP][NP]>(elem_accum_kener_ptr,num_elems);
+  h_PEner     = HostViewUnmanaged<Real*[4][NP][NP]>(elem_accum_pener_ptr,num_elems);
+
+  // Sync the device copy of Q
+  m_Q = ExecViewManaged<Real*[QSIZE_D][NUM_PHYSICAL_LEV][NP][NP]>("Q",num_elems);
+  Kokkos::deep_copy(m_Q,h_Q);
+}
+
+void Diagnostics::update_q(const int np1_qdp, const int np1)
+{
+  // Get simulation params
+  SimulationParams& params = Context::singleton().get_simulation_params();
+  assert(params.params_set);
+
+  // Get hybrid vertical coordinate
+  HybridVCoord& hvcoord = Context::singleton().get_hvcoord();
+  auto hyai_delta = hvcoord.hybrid_ai_delta;
+  auto hybi_delta = hvcoord.hybrid_bi_delta;
+  const Real ps0 = hvcoord.ps0;
+
+  // Get qdp from Tracers and ps_v from Elements
+  Tracers& tracers = Context::singleton().get_tracers();
+  Elements& elements = Context::singleton().get_elements();
+  auto qdp = Kokkos::create_mirror_view(tracers.qdp);
+  auto ps_v = Kokkos::create_mirror_view(elements.m_ps_v);
+  auto Q = m_Q;
+
+  const int num_elems = m_num_elems;
```

Comment:
This kernel will be needed for physics-dynamics coupling, too, because the coupling code needs both Qdp and Q. I recommend these lines be put into a free function externed by Tracers.hpp.

Reply:
Well, then we may as well keep the function inside Tracers...

Reply:
Btw, the code above was buggy and would have not worked on GPU.